pkg/controllers/tagging/tagging_controller.go

```diff
@@ -0,0 +1,248 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package tagging
+
+import (
+	"fmt"
+	v1 "k8s.io/api/core/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	cloudprovider "k8s.io/cloud-provider"
+	opt "k8s.io/cloud-provider-aws/pkg/controllers/options"
+	awsv1 "k8s.io/cloud-provider-aws/pkg/providers/v1"
+	"k8s.io/klog/v2"
+	"time"
+)
+
+// workItem contains the node and an action for that node
+type workItem struct {
+	node   *v1.Node
+	action func(node *v1.Node) error
+}
+
+// TaggingController is the controller implementation for tagging cluster resources.
+// It periodically check for Node events (creating/deleting) to apply appropriate
+// tags to resources.
+type TaggingController struct {
+	nodeInformer coreinformers.NodeInformer
+	kubeClient   clientset.Interface
+	cloud        *awsv1.Cloud
+	workqueue    workqueue.RateLimitingInterface
+	nodesSynced  cache.InformerSynced
+	// Value controlling TaggingController monitoring period, i.e. how often does TaggingController
+	// check node list. This value should be lower than nodeMonitorGracePeriod
+	// set in controller-manager
+	nodeMonitorPeriod time.Duration
+
+	// Representing the user input for tags
+	tags map[string]string
+
+	// Representing the resources to tag
+	resources []string
+}
+
+// NewTaggingController creates a NewTaggingController object
+func NewTaggingController(
+	nodeInformer coreinformers.NodeInformer,
+	kubeClient clientset.Interface,
+	cloud cloudprovider.Interface,
+	nodeMonitorPeriod time.Duration,
+	tags map[string]string,
+	resources []string) (*TaggingController, error) {
+
+	awsCloud, ok := cloud.(*awsv1.Cloud)
+	if !ok {
+		err := fmt.Errorf("tagging controller does not support %v provider", cloud.ProviderName())
+		return nil, err
+	}
+
+	tc := &TaggingController{
+		nodeInformer:      nodeInformer,
+		kubeClient:        kubeClient,
+		cloud:             awsCloud,
+		nodeMonitorPeriod: nodeMonitorPeriod,
+		tags:              tags,
+		resources:         resources,
+		workqueue:         workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "Tagging"),
+		nodesSynced:       nodeInformer.Informer().HasSynced,
+	}
+
+	// Use shared informer to listen to add/update/delete of nodes. Note that any nodes
+	// that exist before tagging controller starts will show up in the update method
+	tc.nodeInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    func(obj interface{}) { tc.enqueueNode(obj, tc.tagNodesResources) },
+		UpdateFunc: func(oldObj, newObj interface{}) { tc.enqueueNode(newObj, tc.tagNodesResources) },
```

Comment:
I don't think this is going to work.  Right now you never remove tags unless the node is deleted, right?  How does a user change an existing tag that they previously set?

Reply:
This is to tag any nodes that exist before the controller were brought up, similar to the [node controller](https://github.com/kubernetes/cloud-provider/blob/master/controllers/node/node_controller.go#L142).

Reply:
Yeah and I just realized you will basically don't delete tags unless nodes are deleted?  Probably partially because we can't recognize tags applied by the controller from user-owned tags...

Reply:
So I guess we should document (maybe you already did) that the controller only deletes tags on a node when the node object gets deleted. It doesn't really expect the tags to change.  And if the tags do change, we just leave the existing tags alone and apply the new ones?

Reply:
Will add it to the controller md

Reply:
@nckturner right now the tags are passed as flags to this controller so we don't really expect them to change. But you have a good point - in case we decide to move these to a config map or make them more configurable, we'll probably need to delete the tags as well.